### PR TITLE
[Perf] Speed up argument processing on Python-side.

### PR DESCRIPTION
**Pull request by duburcqa**
Benchmark for non-batched cpu-based single franka On Genesis zero-copy branch (https://github.com/Genesis-Embodied-AI/Genesis/pull/2011) (219fb52f):

### BEFORE

* field: 5900FPS
* ndarray: 5800FPS

image

### AFTER

* field: 6450FPS
* ndarray: 6250FPS

image

### CODE REVIEW

File: python/gstaichi/lang/kernel_impl.py [hunk 1]
```diff
@@ -390,57 +393,47 @@ def _process_args(
...
+                    if key == arg_meta.name:
+                        if i < num_args:
+                            raise GsTaichiSyntaxError(f"Multiple values for argument '{key}'.")
+                        break
+                else:
+                    raise GsTaichiSyntaxError(f"Unexpected argument '{key}'.")
+    elif num_missing_args:
+        for arg_meta in self.arg_metas_expanded[num_args:]:
+            if fused_args is _ARG_EMPTY:
+                raise GsTaichiSyntaxError(f"Missing argument '{arg_meta.name}'.")
```
Comment (hughperkins) [inline python/gstaichi/lang/kernel_impl.py:438]: I remember that I added all that debug info, otherwise it's really hard to debug stuff. And not just for me, but people who are using gstaichi. To be fair, you are the person who will have to deal with its absence, so ultimately I will defer to your decision on this, and you can always put it back later :) But: are you sure you want to remove this debug info?

File: python/gstaichi/lang/kernel_impl.py [hunk 2]
```diff
@@ -390,57 +393,47 @@ def _process_args(
...
-
-    for key, value in kwargs.items():
-        for i, arg in enumerate(self.arg_metas_expanded):
-            if key == arg.name:
-                if i < len_args:
-                    raise GsTaichiSyntaxError(f"Multiple values for argument '{key}'.")
+    missing_arg_metas = self.arg_metas_expanded[num_args:]
+    num_missing_args = len(missing_arg_metas)
+    fused_args: list[Any] = [*args, *[arg_meta.default for arg_meta in missing_arg_metas]]
+    if kwargs:
```
Comment (hughperkins) [inline python/gstaichi/lang/kernel_impl.py:414]: can you remind me what test(s) we have for this section of code please? (ideally with urls, line numbers, etc).

File: python/gstaichi/lang/kernel_impl.py (hunk 1 above)
Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:438]: Yes, I removed them on purpose, because ultimately I found it more confusing than helpful. This was too verbose and I was struggling understanding what was the issue. In this case, all I need to know is which variable is missing, passed multiple times, or unexpected. Nothing more, nothing less. Same as for any using Python function.

But I'm also ok to restore the original behaviour if you think it is important.

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:438]: No. You are the client. When it's me debugging, I throw in whatever spam I want, all over the place :)

File: python/gstaichi/lang/kernel_impl.py (hunk 2 above)
Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: They are defined in `tests/python/test_argument.py`.

I double checked manually that it is going through all the branches!

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: Actually there is one single branch that is not checked in unit test yet. Let me add it.

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: Fixed! It was actually broken x)

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: could you provide a url pelase, cos I'm really lazy, and can then just click on it, rather than having to go to gstaichi website, and click my way through the hierarchy :P (You'll notice I usually provide you with direct URLs when referring to code I think?)

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: Here it is:
https://github.com/Genesis-Embodied-AI/gstaichi/blob/main/tests/python/test_argument.py

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: Thanks! Hmmm, I'm going to add a ticket to Linear to add SonarQube or smilar to the CI, which makes checking these things much easier.

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: In the absence of Sonarqube, I asked Cursor what it thinks. It says:

Screenshot 2025-11-24 at 10 18 07

Thoughts?

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: * 1.: Maybe, I don't know what this is, and did not modified this part of the code.
* 2.: Same as 1.
* 3.: Not true
* 4.: Not true
* 5.: Not true
* 6.: This is supported and not an edge case. Just the normal code flow. But sure there is no unit test for this. I can add one if you want.
* 7.: Not true.

The code Cursor is referring to does not exist.

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: Yeah, for 1 and 2, seems plausible that it doesnt relate to your change, though I think that 2 refers to python dataclasses, but I feel we've probably got ok-ish coverage for python dataclasses, so doesn't worry me unduly.

Point 3, what do you mean by 'not true'? What test do we have for this currnetly?
- I see a mixture https://github.com/Genesis-Embodied-AI/gstaichi/blob/64195c3dcc43ef82ba5bb7dcfbaaf43a593c4993/tests/python/test_argument.py#L15
- missing kwargs https://github.com/Genesis-Embodied-AI/gstaichi/blob/64195c3dcc43ef82ba5bb7dcfbaaf43a593c4993/tests/python/test_argument.py#L19
- more missing kwargs https://github.com/Genesis-Embodied-AI/gstaichi/blob/64195c3dcc43ef82ba5bb7dcfbaaf43a593c4993/tests/python/test_argument.py#L30
- this one doesnt even have any kernel args at all https://github.com/Genesis-Embodied-AI/gstaichi/blob/64195c3dcc43ef82ba5bb7dcfbaaf43a593c4993/tests/python/test_argument.py#L81
- etc ...

Could you also point to a test case for each for 4,5,6,7 please?

For 'does not exist', it exists in my cloned branch, see screenshot, but sesems to no longer exist in cloud 🙌 
Screenshot 2025-11-24 at 10 54 47

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: Honestly I don't care which test is checking what because I already made sure we are going through all the branches but OK.

* 3.: `test_function_keyword_args`
* 4.: `test_function_keyword_args_duplicate`
* 5.: `test_kernel_args_missing`, `test_kernel_keyword_args_missing`
* 6.: No test for this. I see no point. It is not an edge case at all.
* 7.: We have check for defaults but not all default (`test_function_keyword_args`). This is NOT an edge case either.

I will had a unit test for 6. and 7. even if I don't get the point.

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: Done.

* 6.: `test_function_keyword_args`
* 7.: `test_function_keyword_args`

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: thanks! Yes, like I say, Sonarqube would make this all much easier, more objective. We should set that up soon-ish I feel.

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: (or some equivalent. but I've only ever used Sonarqube, and I know that Sonarqube has a clear ui showing, for the ocde, which new lines are uncovered, etc)

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: sonarqube-covered-test-units

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: > Sonarqube would make this all much easier, more objective. 

I would say more observable. Because my check was objective, I added breakpoint() add the beginning of each branch and deleted them one by one after trigger any of them when running the test suite.

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: Yes, observable to the reviewer. And more repeatable. Icannot see that you have done that, nor can either of us easily re-run your checks after each new commit.

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: This would be indeed much much better, big fan of coverage report.

Comment (duburcqa) [reply python/gstaichi/lang/kernel_impl.py:414]: For the record:
https://github.com/Genesis-Embodied-AI/Genesis/pull/1223

Comment (hughperkins) [reply python/gstaichi/lang/kernel_impl.py:414]: We used to have sonarqube run asynchronously, once everything else had finished running. Since, as you allude to, it's very slow.
